torch/_inductor/scheduler.py

```diff
@@ -806,42 +889,58 @@ def fuse_nodes_once(self):
             - self.score_fusion(): assigns priority to a given fusion
         """
         fused_nodes = set(self.nodes)
+
+        # recent_fusions tracks names of nodes fused during this function call.
+        #
+        # This is required because config.fusion_reorder_loops, when enabled,
+        # causes get_possible_fusions() to return new nodes not in
+        # name_to_fused_node, so we can't assume
+        #   node = self.name_to_fused_node[node.get_first_name()]
+        # will map an unfused nodes back to itself
+        recent_fusions: Set[str] = set()
+
         for node1, node2 in self.get_possible_fusions():
-            node1 = self.name_to_fused_node[node1.get_first_name()]
-            node2 = self.name_to_fused_node[node2.get_first_name()]
+            if node1.get_first_name() in recent_fusions:
+                node1 = self.name_to_fused_node[node1.get_first_name()]
+            if node2.get_first_name() in recent_fusions:
+                node2 = self.name_to_fused_node[node2.get_first_name()]
+
             if self.can_fuse(node1, node2) and not self.will_fusion_create_cycle(
                 node1, node2
             ):
                 node3 = FusedSchedulerNode.fuse(node1, node2)
-                fused_nodes.remove(node1)
-                fused_nodes.remove(node2)
+                fused_nodes.remove(self.name_to_fused_node[node1.get_first_name()])
+                fused_nodes.remove(self.name_to_fused_node[node2.get_first_name()])
                 fused_nodes.add(node3)
                 self.name_to_fused_node.update(
                     {n.get_name(): node3 for n in node3.get_nodes()}
                 )
+
+                recent_fusions.add(node1.get_first_name())
+                recent_fusions.add(node2.get_first_name())
+
         self.nodes = sorted(fused_nodes, key=lambda x: x.min_order)
         self.topological_sort_schedule()
 
     def get_possible_fusions(self):
         """
         Helper to find all legal fusion opportunities, sorted by self.score_fusion()
         """
-        possible_fusions = []
+        possible_fusions = set()
```

Comment:
This might make things non-deterministic since set iteration order can change run-to-run.

Reply:
The fusions are sorted before being returned, and the set itself is only used for `.add` and `in`.

Reply:
I believe they are sorted by score_fusion, which can return duplicates.

Reply:
Fair point, we could add a tie breaker to `score_fusion` like `id(node1)`?

Reply:
`id()` is also non-determinstic.  (It is the internal memory address.)